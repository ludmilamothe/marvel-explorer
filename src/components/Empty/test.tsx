import { screen } from '@testing-library/react';
import { renderWithProviders } from 'utils/test/helpers';

import Empty from '.';

const datas = {
  imagePath: '/images/no-search.svg',
  imageAlt:
    'A magnifying glass with sad face and with some bubbles in the background',
  children: 'Some text',
};

describe('<Empty />', () => {
  it('should render the empty correct', () => {
    renderWithProviders(<Empty {...datas} />);

    const image = screen.getByRole('img', { name: datas.imageAlt });
    const children = screen.getByText(datas.children);

    expect(children).toBeInTheDocument();

    expect(image).toBeInTheDocument();
    expect(image).toHaveAttribute('src', datas.imagePath);
  });

  it('should render the empty styles correct', () => {
    const { container } = renderWithProviders(<Empty {...datas} />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
