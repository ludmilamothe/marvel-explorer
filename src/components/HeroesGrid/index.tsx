import styled, { css } from 'styled-components';
import media from 'styled-media-query';

export type HeroesGridProps = {
  applyAutoFit?: boolean;
};

const heroesGridModifiers = {
  applyAutoFit: () => css`
    grid-template-columns: repeat(auto-fit, minmax(12.8rem, 1fr));

    ${media.greaterThan('small')`
      grid-template-columns: repeat(auto-fit, minmax(16rem, 1fr));
    `}
  `,
};

const HeroesGrid = styled.div<HeroesGridProps>`
  ${({ theme, applyAutoFit = false }) => css`
    width: 100%;
    display: grid;
    gap: ${theme.spacings.xsmall};
    grid-template-columns: repeat(2, minmax(12.8rem, 1fr));

    ${media.greaterThan('small')`
      gap: ${theme.spacings.small};
      grid-template-columns: repeat(5, minmax(16rem, 1fr));
    `}

    ${applyAutoFit && heroesGridModifiers.applyAutoFit}
  `}
`;

export default HeroesGrid;
