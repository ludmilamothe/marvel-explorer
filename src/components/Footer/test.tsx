import { screen } from '@testing-library/react';
import { renderWithTheme } from 'utils/test/helpers';

import Footer from '.';

describe('<Footer />', () => {
  it('should render the footer content correct', () => {
    renderWithTheme(<Footer />);

    const linkText = 'Mauro de Souza';
    const copy = 'Data provided by Marvel. Â© 2020 MARVEL';

    const linkElement = screen.getByRole('link', { name: linkText });

    expect(linkElement).toHaveAttribute(
      'href',
      'https://www.linkedin.com/in/maurodesouzaa/'
    );

    expect(linkElement).toHaveAttribute('target', '_blank');

    expect(screen.getByText(copy)).toBeInTheDocument();
  });

  it('should render the footer styles correct', () => {
    const { container } = renderWithTheme(<Footer />);

    const footer = container.firstChild;

    expect(footer).toMatchSnapshot();
  });
});
