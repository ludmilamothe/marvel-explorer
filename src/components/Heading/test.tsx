import { screen } from '@testing-library/react';
import { renderWithTheme } from 'utils/test/helpers';

import theme from 'styles/theme/default';
import Heading from '.';

describe('<Heading />', () => {
  it('should render the heading with children', () => {
    const text = 'Just a text to test';

    renderWithTheme(<Heading>{text}</Heading>);

    const heading = screen.getByRole('heading', { name: text });

    expect(heading).toBeInTheDocument();

    expect(heading).toMatchInlineSnapshot(`
      .c0 {
        color: #FFF;
        font-weight: 900;
        font-size: 2.4rem;
        max-width: 55rem;
        text-align: center;
      }

      @media (min-width:768px) {
        .c0 {
          font-size: 4.0rem;
        }
      }

      <h1
        class="c0"
      >
        Just a text to test
      </h1>
    `);
  });

  it('should render the heading with correct styles', () => {
    const text = 'Just a text to test';

    renderWithTheme(<Heading>{text}</Heading>);

    const heading = screen.getByRole('heading', { name: text });

    expect(heading).toHaveStyle({
      fontSize: theme.font.sizes.large,
      fontWeight: theme.font.xbold,
      color: theme.colors.white,
    });

    expect(heading).toHaveStyleRule('font-size', theme.font.sizes.xlarge, {
      media: '(min-width: 768px)',
    });
  });
});
