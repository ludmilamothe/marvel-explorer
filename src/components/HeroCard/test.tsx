/* eslint-disable @typescript-eslint/no-var-requires */

import { fireEvent, screen } from '@testing-library/react';
import { renderWithProviders } from 'utils/test/helpers';

import { storageKey } from 'hooks/useFavoriteHeroes';
import HeroCard from '.';

// eslint-disable-next-line @typescript-eslint/no-var-requires
const useRouter = jest.spyOn(require('next/router'), 'useRouter');
const push = jest.fn();

useRouter.mockImplementation(() => ({
  push,
}));

const heroData = {
  heroId: 123,
  heroName: 'Iron man',
  heroThumbnail:
    'http://i.annihil.us/u/prod/marvel/i/mg/9/c0/527bb7b37ff55.jpg',
};

describe('<HeroCard />', () => {
  beforeEach(() => {
    window.localStorage.clear();
  });

  it('should render the hero card correct', () => {
    renderWithProviders(<HeroCard {...heroData} />);

    expect(
      screen.getByRole('img', { name: heroData.heroName })
    ).toBeInTheDocument();

    expect(
      screen.getByRole('heading', { name: heroData.heroName })
    ).toBeInTheDocument();

    expect(screen.getByLabelText(/add hero to favorites/i)).toBeInTheDocument();
  });

  it('should render the hero card styles correct', () => {
    const { container } = renderWithProviders(<HeroCard {...heroData} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should add hero to favorites', () => {
    renderWithProviders(<HeroCard {...heroData} />);

    const addToFavoriteElement = screen.getByLabelText(
      /add hero to favorites/i
    );

    expect(addToFavoriteElement).toBeInTheDocument();
    fireEvent.click(addToFavoriteElement);

    const removeFromFavoriteElement = screen.getByLabelText(
      /remove hero to favorites/i
    );

    expect(removeFromFavoriteElement).toBeInTheDocument();
    expect(addToFavoriteElement).not.toBeInTheDocument();

    const storagedHero = JSON.parse(window.localStorage.getItem(storageKey)!);

    expect(storagedHero).toEqual([heroData]);
  });

  it('should remove hero to favorites', () => {
    window.localStorage.setItem(storageKey, JSON.stringify([heroData]));

    renderWithProviders(<HeroCard {...heroData} />);

    const removeFromFavoriteElement = screen.getByLabelText(
      /remove hero to favorites/i
    );

    expect(removeFromFavoriteElement).toBeInTheDocument();
    fireEvent.click(removeFromFavoriteElement);

    const addToFavoriteElement = screen.getByLabelText(
      /add hero to favorites/i
    );

    expect(removeFromFavoriteElement).not.toBeInTheDocument();
    expect(addToFavoriteElement).toBeInTheDocument();

    const storagedHero = JSON.parse(window.localStorage.getItem(storageKey)!);
    expect(storagedHero).toEqual([]);
  });

  it('should navigate to hero page', () => {
    const { container } = renderWithProviders(<HeroCard {...heroData} />);

    const heroCard = container.firstChild!;
    fireEvent.click(heroCard);

    expect(push).toBeCalledTimes(1);
    expect(push).toBeCalledWith(`/characters/${heroData.heroId}`);
  });
});
