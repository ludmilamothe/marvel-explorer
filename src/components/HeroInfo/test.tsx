import { fireEvent, screen } from '@testing-library/react';
import { renderWithProviders } from 'utils/test/helpers';

import HeroInfo from '.';

const props = {
  id: 123,
  name: 'iron man',
  description:
    'Inventor Tony Stark applies his genius for high-tech solutions to problems as Iron Man, the armored Avenger.',
  wikiLink: 'https://google.com',
  comics: 123,
  series: 1234,
  stories: 12345,
};

describe('<HeroInfo />', () => {
  it('should render the HeroInfo correct', () => {
    renderWithProviders(<HeroInfo {...props} />);

    const link = screen.getByRole('link', { name: /see more/i });
    expect(screen.getByText(props.name)).toBeInTheDocument();

    expect(screen.getByText(props.description)).toBeInTheDocument();
    expect(screen.getByText(props.comics)).toBeInTheDocument();

    expect(screen.getByText(props.series)).toBeInTheDocument();
    expect(screen.getByText(props.stories)).toBeInTheDocument();

    expect(link).toBeInTheDocument();
    expect(link).toHaveAttribute('href', props.wikiLink);
  });

  it('should alter icon between add/remove favorite', () => {
    renderWithProviders(<HeroInfo {...props} />);

    let addToFavoriteElement = screen.getByLabelText(/add hero to favorites/i);

    expect(addToFavoriteElement).toBeInTheDocument();
    fireEvent.click(addToFavoriteElement);

    const removeFromFavoriteElement = screen.getByLabelText(
      /remove hero to favorites/i
    );

    expect(removeFromFavoriteElement).toBeInTheDocument();
    expect(addToFavoriteElement).not.toBeInTheDocument();

    fireEvent.click(removeFromFavoriteElement);

    addToFavoriteElement = screen.getByLabelText(/add hero to favorites/i);

    expect(removeFromFavoriteElement).not.toBeInTheDocument();
    expect(addToFavoriteElement).toBeInTheDocument();
  });
});
