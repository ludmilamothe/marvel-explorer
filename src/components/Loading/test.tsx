import { renderWithProviders } from 'utils/test/helpers';

import Loading from '.';

describe('<Loading />', () => {
  it('should render the Loading', () => {
    const { container } = renderWithProviders(<Loading />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
