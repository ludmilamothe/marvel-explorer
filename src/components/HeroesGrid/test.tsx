import { renderWithProviders } from 'utils/test/helpers';

import HeroesGrid from '.';

describe('<HeroesGrid />', () => {
  it('should render the HeroesGrid styles correct', () => {
    const { container } = renderWithProviders(<HeroesGrid />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should aplly auto fit styles when the prop is passed', () => {
    const { container } = renderWithProviders(<HeroesGrid applyAutoFit />);

    const component = container.firstChild;

    expect(component).toHaveStyle({
      'grid-template-columns': 'repeat(auto-fit,minmax(12.8rem,1fr))',
    });

    expect(component).toHaveStyleRule(
      'grid-template-columns',
      'repeat(auto-fit,minmax(16rem,1fr))',
      {
        media: '(min-width: 450px)',
      }
    );
  });
});
