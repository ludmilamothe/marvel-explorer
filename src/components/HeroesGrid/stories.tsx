import { Meta, Story } from '@storybook/react/types-6-0';

import HeroCard from '../HeroCard';

import HeroesGrid, { HeroesGridProps } from '.';

type HeroesGridStoryProps = HeroesGridProps & {
  renderHeroCards: number;
};

const props = {
  heroId: 123,
  heroName: 'Iron man',
  heroThumbnail:
    'http://i.annihil.us/u/prod/marvel/i/mg/9/c0/527bb7b37ff55.jpg',
};

export default {
  title: 'HeroesGrid',
  component: HeroesGrid,

  argTypes: {
    theme: { table: { disable: true } },
    ref: { table: { disable: true } },
    as: { table: { disable: true } },
    forwardedAs: { table: { disable: true } },
  },

  args: {
    renderHeroCards: 1,
  },
} as Meta<HeroesGridStoryProps>;

const createArray = (lenght: number): string[] => new Array(lenght).fill('');

export const Basic: Story<HeroesGridStoryProps> = args => {
  const { renderHeroCards, ...rest } = args;

  return (
    <div style={{ width: '100%', maxWidth: '1040px' }}>
      <HeroesGrid {...rest}>
        {createArray(renderHeroCards).map((_, index) => (
          <HeroCard key={props.heroId * (index + 1)} {...props} />
        ))}
      </HeroesGrid>
    </div>
  );
};
