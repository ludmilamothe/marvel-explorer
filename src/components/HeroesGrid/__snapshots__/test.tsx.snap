// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HeroesGrid /> should render the HeroesGrid styles correct 1`] = `
.c0 {
  width: 100%;
  display: grid;
  gap: 1.2rem;
  grid-template-columns: repeat(2,minmax(12.8rem,1fr));
}

@media (min-width:450px) {
  .c0 {
    gap: 2.4rem;
    grid-template-columns: repeat(5,minmax(16rem,1fr));
  }
}

<div
  class="c0"
/>
`;
